### PR TITLE
✨ Add --exclude-packges Flag to osv-scanner CLI

### PR DESCRIPTION
**Pull request by toddtee**
This PR adds an `--exclude-packages` flag to the osv-scanner CLI. This flag allows users to specify a list of regex patterns to which matches of collected package names will be matched against.

If the packages are found to match the regex patterns, they will be excluded from the query to the OSV database.

Exclusions work for both lockfile and SBOM package definitions.

Excluded-packages flag will address issue #151.

### CODE REVIEW

File: pkg/exclusions/exclusions.go
Comment (G-Rath) [inline pkg/exclusions/exclusions.go:87]: Regexp compiling should be done in `cachedregexp` - currently that only has `MustCompile` but this is a reason for `Compile`, and is extra nice because we know its going to get `Compile`'d at least twice

Comment (toddtee) [reply pkg/exclusions/exclusions.go:87]: Hey mate, just going to play back my understanding to make sure I fully understand your suggestion.

Thanks for pointing out the `cachedregex` package; definitely makes sense to move the regex handling to that package. So for that reason:

- I will also create a `Compile` function for the `cachedregex` package which can be called by the `checkForRegexMatch` function:
```
// Compile takes a regex pattern and returns a list of Regexp objects.
func Compile(pattern string) (*regexp.Regexp, error) {

	re, err := regexp.Compile(pattern)
	if err != nil {
		return nil, fmt.Errorf("invalid regex pattern: %s", pattern)
	}
	return re, nil
}
```

-  I will move `exclusions.isValidRegex` to `cachedregex.IsValidRegex`, which will also make use of the new `Compile` function.

```
// IsValidRegex checks if a regex pattern is valid.
func IsValidRegex(pattern string) bool {
	_, err := Compile(pattern)
	return err == nil
}
```

I have also updated the `parseExclusions` function to utilise the `cachedregexp.MustCompile` fucntionality as that function doesn't care about errors?

```
// parseExclusions parses a list of exclusion patterns and returns a list of Regexp objects.
func parseExclusions(patterns []string) ([]*regexp.Regexp, error) {
	exclusionRegexes := make([]*regexp.Regexp, len(patterns))
	for i, pattern := range patterns {
		re := cachedregexp.MustCompile(pattern)
		exclusionRegexes[i] = re
	}

	return exclusionRegexes, nil
}
```

Am I roughly in the ballpark or way off?

Thanks for your time.

Comment (G-Rath) [reply pkg/exclusions/exclusions.go:87]: Just create the `Compile` function in `cachedregexp` - `isValidRegex` should remain where it is

Comment (toddtee) [reply pkg/exclusions/exclusions.go]: Thanks @G-Rath, updated now. 👌

File: docs/usage.md
Comment (G-Rath) [inline docs/usage.md:150]: nit: I feel like we should be mentioning the flag at the end of this sentence:

```suggestion
OSV database, you can pass a list of regex patterns with the `--exclusions` flag. The `osv-scanner` will
```

File: docs/usage.md
Comment (G-Rath) [inline docs/usage.md:158]: nit: I think it would be better to use single quotes as (at least in Bash) they don't do interpolation and the nature of regex syntax makes me paranoid of that happening by mistake 😅 

```suggestion
osv-scanner --exclusions='^example.com/.*,^react$' <rest-of-cli-interaction>
```

(it might even be worth adding a warning about that?)

File: pkg/exclusions/exclusions.go
Comment (G-Rath) [inline pkg/exclusions/exclusions.go:15]: nit: while I like comments, I feel like a few of them in this file are redundant

e.g. I think it should be pretty obvious given the return type that the updated lockfile is what is being returned, and that a private function called `isRegexValid` checks if strings are valid regexps...

Would you mind doing re-check with that thinking in your head and seeing if there are any comments you're comfortable with dropping? I know this can be very subjective so happy if you come back saying you think we need all these comments 🙂

File: pkg/osvscanner/osvscanner.go
Comment (G-Rath) [inline pkg/osvscanner/osvscanner.go:41]: I'm softly leaning towards saying we should be taking a slice of regexps here because that is the most accurate (and thus least-error-prone) representation of the data; the reason I'm not completely sold is that I expect a number of library consumers are probably taking user input aka dynamic regexps, but for us it should mean the library doesn't have to handle "what if the string isn't a valid regexp"...

This would be a reason to keep `exclusions` as a public package as then downstream users could just pull that in like the CLI does rather than have to implement their own parser (then again they could also just copy and paste, which is the Go way 🤷)

File: pkg/exclusions/exclusions.go
Comment (G-Rath) [inline pkg/exclusions/exclusions.go:1]: This should be `internal` by default since public packages have to follow semver

**Review by G-Rath — CHANGES_REQUESTED** [review]
This is on the right track - I've done a general review but not gone over every line yet as I think the next main question is about if the library should be accepting a slice of regexp or strings, as changing that will probably mean other code can be changed which'd make other parts of my review moot 😅

File: docs/usage.md
Comment (toddtee) [reply docs/usage.md:159]: Bash in general makes me paranoid, so I get it... 🤣

File: docs/usage.md
Comment (hayleycd) [inline docs/usage.md:151]:
```suggestion
flag. OSV-Scanner will check for packages matching the specified regex
```
This phrasing is more consistent with our documentation.

Comment (toddtee) [reply docs/usage.md:151]: Sure thing!

File: pkg/exclusions/exclusions.go
Comment (toddtee) [reply pkg/exclusions/exclusions.go:15]: Completely agree. Comments are good, but superfluous commenting can reduce readability in some cases. I do believe it is good practice to leave comments on public functions though.

**Review by oliverchang — COMMENTED** [review]
Thanks for this! Our team has considered this in the past as a config file rule (https://google.github.io/osv-scanner/configuration/).

Could we understand if a config file would also work for your use case? Or are there reasons why you'd prefer a CLI flag?

File: docs/usage.md
Comment (hayleycd) [inline docs/usage.md:147]: Thank you for updating the documentation for the --exclusions flag. I appreciate you including this as part of your PR. 

Our GitHub Page theme automatically wraps text. Each paragraph should be included in a single line. This makes future changes easier and is more consistent with the majority of the documentation (though there are a few places where I need to fix this).

**Comment by hayleycd** [comment]
Docs look good with a couple of comments.

File: docs/usage.md
Comment (toddtee) [reply docs/usage.md]: Updated.

File: docs/usage.md
Comment (toddtee) [reply docs/usage.md:147]: Hi @hayleycd did you need me to update the paragraphs to be a single line to fit in with your Github Pages theme? Or it is fine to leave the paragraph wrapped?

**Comment by toddtee** [comment]
Hi @oliverchang, thanks for checking out my PR!

> Could we understand if a config file would also work for your use case? Or are there reasons why you'd prefer a CLI flag?

Ideally, I think having the `exclusions` parameter both as a CLI flag and a configuration parameter would be ideal. 

I like the idea of having the CLI flag as I personally believe that CLI tools should have good levels of interactivity for a human user. For example, if I as a developer want to quickly update a project and I have a particular package that is giving me issues that I am blocked on updating or don't want to query for whatever reason, it is nice to have the flag as a quick option to interact with the tool as I need it. Or I could even use the flag to test the scan results for when I am developing the regex patterns that will eventually end up in the configuration file! 🤣 

However... let's now talk automation, not human interaction. If I incorporate the `osv-scanner` as part of my project's CI/CD pipeline or automated security checks, having the flag available is good, however having a configuration parameter is better. That way, maintainers of the project can configure what exclusions they like and perhaps even have comments in the config file as to why they are there. I see the configuration parameter for more permanent and/or governing exclusion purposes.

**TL;DR, I see the CLI flag for quick developer use for once-offs or short term purposes and the configuration file more for automated processes and more permanent solutions.** 

I don't know how challenging incorporating the configuration file into my current offering would be in the short term...

However, as I am a supporter of one PR including one change only. I would propose that this PR maintain adding the functionality of a CLI flag exclusively, and perhaps create a related Issue and follow-up PR to add exclusions as a configuration option as well. 

The benefit of this is the tool can still offer the ability to exclude packages from the query in the short term and deliver value to those that need it. A configuration parameter will extend that functionality offering further.

I hope that answers your question 🙏

File: docs/usage.md
Comment (hayleycd) [reply docs/usage.md:147]: I added a commit. Docs are good to go now.

File: pkg/exclusions/exclusions.go
Comment (toddtee) [reply pkg/exclusions/exclusions.go:1]: Yep agreed, ain't no one got time for that.

File: docs/usage.md
Comment (toddtee) [reply docs/usage.md:147]: Thanks @hayleycd, I have squashed this into the main commit.

File: pkg/osvscanner/osvscanner.go
Comment (toddtee) [reply pkg/osvscanner/osvscanner.go:41]: Hey @G-Rath, given in my brain, the user was always going to be input from the user, I naturally just assumed it was going to be a list of string. If it were going to be regexp, we would still need to take the list of string from the flag and parse them as regexp _before_ we assign them to the ScannerActions struct? 🤔

Comment (G-Rath) [reply pkg/osvscanner/osvscanner.go:41]: What you've said is correct but the code base is made up of both the CLI and the library - and it's the CLI that handles the fact the input is coming in as a string (which is because that's just how data comes in from terminals)

The library on the other hand can just accept regexp so it doesn't need to be responsible for handling the edge cases that come with parsing strings into regexp

Comment (toddtee) [reply pkg/osvscanner/osvscanner.go:41]: Thanks @G-Rath that makes sense. Let me shuffle some things and see what happens. 🔀

Comment (toddtee) [reply pkg/osvscanner/osvscanner.go]: Updated @G-Rath.

The `ScannerActions` struct now has the `Exclusions` field as a `[]*regex.regexp`.

**Comment by oliverchang** [comment]
> Hi @oliverchang, thanks for checking out my PR!
> 
> > Could we understand if a config file would also work for your use case? Or are there reasons why you'd prefer a CLI flag?
> 
> Ideally, I think having the `exclusions` parameter both as a CLI flag and a configuration parameter would be ideal.
> 
> I like the idea of having the CLI flag as I personally believe that CLI tools should have good levels of interactivity for a human user. For example, if I as a developer want to quickly update a project and I have a particular package that is giving me issues that I am blocked on updating or don't want to query for whatever reason, it is nice to have the flag as a quick option to interact with the tool as I need it. Or I could even use the flag to test the scan results for when I am developing the regex patterns that will eventually end up in the configuration file! 🤣

Thanks, that makes sense. We've generalised shied away from having multiple ways to achieve the same thing out of a desire to keep the CLI simple to understand, but having a way to quickly iterate/test does make sense. 

Could we perhaps rename the flag to `--exclude-packages` though? `--exclude` is a little ambiguous on its own. 

@another-rex @G-Rath Do you think there's value in maintaining this as a CLI flag? 

> 
> However... let's now talk automation, not human interaction. If I incorporate the `osv-scanner` as part of my project's CI/CD pipeline or automated security checks, having the flag available is good, however having a configuration parameter is better. That way, maintainers of the project can configure what exclusions they like and perhaps even have comments in the config file as to why they are there. I see the configuration parameter for more permanent and/or governing exclusion purposes.

Big +1 to this. Having this stored as the repo would help maintain state around exclusions and reduce duplicate triage efforts for the project.

> 
> **TL;DR, I see the CLI flag for quick developer use for once-offs or short term purposes and the configuration file more for automated processes and more permanent solutions.**
> 
> I don't know how challenging incorporating the configuration file into my current offering would be in the short term...
> 
> However, as I am a supporter of one PR including one change only. I would propose that this PR maintain adding the functionality of a CLI flag exclusively, and perhaps create a related Issue and follow-up PR to add exclusions as a configuration option as well.

Yep! I've filed https://github.com/google/osv-scanner/issues/670 for this. 
> 
> The benefit of this is the tool can still offer the ability to exclude packages from the query in the short term and deliver value to those that need it. A configuration parameter will extend that functionality offering further.
> 
> I hope that answers your question 🙏

**Comment by toddtee** [comment]
@oliverchang, I think `--excluded-packages` is better as it is more explicit to what the functionality actually is. I have updated the flag and documentations to reflect this. 👍

**Comment by G-Rath** [comment]
> Do you think there's value in maintaining this as a CLI flag?

I'm not sure - I'm not getting super strong "omg yes this'd be great" vibes from my gut, but that doesn't meant it still wouldn't be useful.

However...

>  For example, if I as a developer want to quickly update a project and I have a particular package that is giving me issues that I am blocked on updating or don't want to query for whatever reason, it is nice to have the flag as a quick option to interact with the tool as I need it

I'm not sure that is as useful as you think: since we're agreed that it's best to capture this in config long-term, in this situation you'd be running the scanner locally in most cases, meaning you could just mentally ignore the results too - I know that is not as nice as having them omitted but not sure if its "support a new flag forever" nice.

> I could even use the flag to test the scan results for when I am developing the regex patterns that will eventually end up in the configuration file!

That does seem like a useful reason to have the flag, though I don't think it would be that much overhead to just be using the config file; that would also mean less risk that a specific shell might get confused by the regexp.

---

I don't think this should be taken as "never will we ever have this flag", but it feels like it would probably be best to start life via the config file

**Pull request closed without merging**
